{tools}[foss/2023a] ICON

jhgoebbert opened the pull request:
This pull-request adds easyconfig files for the software ICON ( https://gitlab.dkrz.de/icon/icon-model ) which just recently got announced as open-source.
ICON is a flexible, scalable, high-performance modelling framework for weather, climate and environmental prediction that provides actionable information for society and advances our understanding of the Earth's climate system and a candidate for the first European Exascale supercomputer JUPITER.
More on ICON can be found here https://icon-model.org

jhgoebbert commented:
Hello @boegel 
Hello @SebastianAchilles 

the bot is complaining about the following lines:
```
# get the local cuda_compute_capabilities
local_gpu_cc = os.environ["EASYBUILD_CUDA_COMPUTE_CAPABILITIES"].split(',')[0].replace('.', '')
```
This works fine at our side at JSC - and somehow it looks quite elegant to me (no need for an extra easyblock).
Would that be an approach you can support to automatically get the `gpu_ cc` for the `-gpu=cc%s` flag of the NVHPC compiler?

jhgoebbert commented:
I removed the nvhpc 24.1 version of ICON from this pull request to avoid the CI-issue for now.

Flamefire reviewed (commented):
> ```
> # get the local cuda_compute_capabilities
> local_gpu_cc = os.environ["EASYBUILD_CUDA_COMPUTE_CAPABILITIES"].split(',')[0].replace('.', '')
> ```
> 
> This works fine at our side at JSC - and somehow it looks quite elegant to me (no need for an extra easyblock). Would that be an approach you can support to automatically get the `gpu_ cc` for the `-gpu=cc%s` flag of the NVHPC compiler?

Yes it is elegant. The problem is that this won't work if you specify it via cmdline or in the config file. Also it only uses the first one, and the order might be kinda random. There is `%(cuda_sm_comma_sep)s` which would expand to e.g. `sm_61,sm_80`    
So it looks like we need another template here. Like `%(cuda_highest_cc)s` or so.

Anyway: Very nice, I'll forward this to a colleague working with ICON here (ZIH)

jhgoebbert commented on easybuild/easyconfigs/i/ICON/ICON-2024.01-foss-2023a.eb:
```diff
@@ -0,0 +1,244 @@
+        'keep_git_dir': True,
+    },
+    'filename': 'v%s.tar.gz' % version,
+}]
```
Thank you for the hint. I use this explicit tar.gz now. It includes all external submodules.
https://gitlab.dkrz.de/icon/icon-model/-/archive/release-2024.01-public/icon-model-release-2024.01-public.tar.gz

jhgoebbert commented on easybuild/easyconfigs/i/ICON/ICON-2024.01-foss-2023a.eb:
```diff
@@ -0,0 +1,244 @@
+]
+
+preconfigopts = (
+    # Overwrite externals YAC with newer version 
```
I have removed this section as it is only required for NVHPC 24.1.  
(and converted it to a small patch for the nvhpc-ebconfig file which is not part of this pull request)

jhgoebbert commented on easybuild/easyconfigs/i/ICON/ICON-2024.01-foss-2023a.eb:
```diff
@@ -0,0 +1,244 @@
+    'ICON_CFLAGS="-O3" '
+    'ICON_BUNDLED_CFLAGS="-O2" '
+    'CPPFLAGS="'
+    '    -I${EBROOTHDF5}/include '
```
Oh, yes, sure. I removed this lines and just left a comment.

jhgoebbert commented on easybuild/easyconfigs/i/ICON/ICON-2024.01-foss-2023a.eb:
```diff
@@ -0,0 +1,244 @@
+    '" '
+    # Linker
+    'LDFLAGS="'
+    '    -L${EBROOTZLIB}/lib '
```
Yes, same as above.

jhgoebbert commented on easybuild/easyconfigs/l/libfyaml/libfyaml-0.9-GCCcore-12.3.0.eb:
```diff
@@ -0,0 +1,40 @@
+# This easyconfig was created by the BEAR Software team at the University of Birmingham.
```
Ok, I removed that one.

jhgoebbert commented:
> So it looks like we need another template here. Like `%(cuda_highest_cc)s` or so.

It would be great to have such a template. For this pull-request it is not yet needed. But I would like to incude it into ICON's ebconfig file for NVHPC-24.1. Another approach could have been
```
local_gpu_cc_str = '%(cuda_compute_capabilities)s'
local_gpu_cc = local_gpu_cc_str.split(',')[0].replace('.', '')
```
But it seems as if `.split(',')[0].replace('.', '')` cannot be added to a eb-variable - at least I did not get it to work.

> Anyway: Very nice, I'll forward this to a colleague working with ICON here (ZIH)

Thank you for forwarding.

SebastianAchilles commented on easybuild/easyconfigs/i/ICON/ICON-2024.01-foss-2023a.eb:
```diff
@@ -0,0 +1,217 @@
+    ('ecCodes', '2.31.0'),
+    ('netCDF', '4.9.2'),
+    ('netCDF-Fortran', '4.6.1'),
+    ('HDF5', '1.14.3'),
```
In `foss/2023a`, this HDF5 version is used:
```suggestion
    ('HDF5', '1.14.0'),
```

jhgoebbert replied:
Thanks. I fixed it.

SebastianAchilles commented:
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
zen2-rockylinux-89 - Linux Rocky Linux 8.9, x86_64, AMD EPYC 7452 32-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/SebastianAchilles/48809867a3c71e92792b60424fb8171f for a full test report.

Flamefire commented:
> > So it looks like we need another template here. Like `%(cuda_highest_cc)s` or so.
> 
> It would be great to have such a template. For this pull-request it is not yet needed. But I would like to incude it into ICON's ebconfig file for NVHPC-24.1. Another approach could have been
> 
> ```
> local_gpu_cc_str = '%(cuda_compute_capabilities)s'
> local_gpu_cc = local_gpu_cc_str.split(',')[0].replace('.', '')
> ```
> 
> But it seems as if `.split(',')[0].replace('.', '')` cannot be added to a eb-variable - at least I did not get it to work.

This is a bit more involved. What happens there is literally `local_gpu_cc = '%(cuda_compute_capabilities)s'.split(',')[0].replace('.', '')` which ends up as `local_gpu_cc = '%(cuda_compute_capabilities)s'`.   
The point is that this isn't a variable: It is just a format string. At the point of use the equivalent of `run(configopts % {'cuda_compute_capabilities': '6.1,8.0', 'other_template': 'value'})` is executed internally.

Also those templates are special because they are based on the values of the easyconfig itself, so `%(cuda_compute_capabilities)s` is not (much) different to `%(namelower)s`: The template is set **after** the easyconfig is parsed (interpreted as a plain Python file in a local context) so you cannot operate on the values while the file is being parsed. You could maybe do something like `local_gpu_cc = '$(echo "%(cuda_compute_capabilities)s" | awk -F, \'{print $1;}\' | sed "s/.//")'` to operate on it in the shell

Side note for `cuda_compute_capabilities`: The template gets set to the value of that variable in the EC defaulting to the config value (from env/cmdline/config) if not set in the EC. This is actually very useful: We have a hook that sets the EC value based on the current cluster arch and CUDA version used to avoid trying to build for a high SM-arch but not too high for the NVCC compiler.

jhgoebbert commented:
Thank you for the background details. This makes it easier to understand what is possible.  
For now I will then operate with the shell for `local_gpu_cc` to avoid `EASYBUILD_CUDA_COMPUTE_CAPABILITIES` in the ebconfig of ICON for nvhpc.

SebastianAchilles commented:
@boegelbot please test @ jsc-zen3

SebastianAchilles commented:
@boegelbot please test @ generoso

jhgoebbert commented:
Any chance this could be merged?

SebastianAchilles reviewed (approved):
lgtm

SebastianAchilles commented:
Going in, thanks @jhgoebbert!

SebastianAchilles commented:
Going in, thanks @jhgoebbert!
